### PR TITLE
Add shimscript arg to cask DSL

### PR DESCRIPTION
**Pull request by FnControlOption**
- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Sometimes, symlinking doesn't work on a cask binary, so a shim script must be used instead.

```rb
shimscript = "#{staged_path}/tiled.wrapper.sh"
binary shimscript, target: "tiled"

preflight do
  File.write shimscript, <<~EOS
    #!/bin/bash
    exec '#{appdir}/Tiled.app/Contents/MacOS/Tiled' "$@"
  EOS
end
```

can be reduced to

```rb
binary "#{appdir}/Tiled.app/Contents/MacOS/Tiled",
       shimscript: "tiled.wrapper.sh",
       target:     "tiled"
```

Also see: https://github.com/Homebrew/homebrew-cask/issues/18809

### CODE REVIEW

**Review by carlocab — APPROVED** [review]
Makes sense to me, and I think this should cover nearly every use of `shimscript` in Cask.

It would be good to confirm the above.

**Comment by FnControlOption** [comment]
30 casks use shim scripts, 7 of those are special:

```rb
# android-ndk
shimscript = "#{staged_path}/ndk_exec.sh"
preflight do
  build = File.read("#{staged_path}/source.properties").match(/(?<=Pkg.Revision\s=\s\d\d.\d.)\d+/)
  FileUtils.ln_sf("#{staged_path}/AndroidNDK#{build}.app/Contents/NDK", "#{HOMEBREW_PREFIX}/share/android-ndk")

  File.write shimscript, <<~EOS
    #!/bin/bash
    readonly executable="#{staged_path}/AndroidNDK#{build}.app/Contents/NDK/$(basename ${0})"
    test -f "${executable}" && exec "${executable}" "${@}"
  EOS
end

%w[
  ndk-build
  ndk-depends
  ndk-gdb
  ndk-stack
  ndk-which
].each { |link_name| binary shimscript, target: link_name }
```

```rb
# dynamodb-local
preflight do
  File.write shimscript, <<~EOS
    #!/bin/sh
    cd "$(dirname "$(readlink -n "${0}")")" && \
      java -Djava.library.path='./DynamoDBLocal_lib' -jar 'DynamoDBLocal.jar' "$@"
  EOS
end
```

```rb
# gog
preflight do
  File.write shimscript, <<~EOS
    #!/bin/sh
    cd '#{staged_path}/gogs' && ./gogs "$@"
  EOS
end
```

```rb
# minecraft-server
config_dir = HOMEBREW_PREFIX.join("etc", "minecraft-server")

preflight do
  FileUtils.mkdir_p config_dir

  File.write shimscript, <<~EOS
    #!/bin/sh
    cd '#{config_dir}' && \
      exec /usr/bin/java ${@:--Xms1024M -Xmx1024M} -jar '#{staged_path}/server.jar' nogui
  EOS
end
```

```rb
# p4v
p4_wrapper = "#{staged_path}/p4.wrapper.sh"
binary p4_wrapper, target: "p4v"
binary p4_wrapper, target: "p4admin"
binary p4_wrapper, target: "p4merge"

preflight do
  File.write p4_wrapper, <<~EOS
    #!/bin/bash
    set -euo pipefail
    COMMAND=$(basename "$0")
    if [[ "$COMMAND" == "p4merge" ]]; then
      exec "#{appdir}/${COMMAND}.app/Contents/Resources/launch${COMMAND}" "$@" 2> /dev/null
    else
      exec "#{appdir}/${COMMAND}.app/Contents/MacOS/${COMMAND}" "$@" 2> /dev/null
    fi
  EOS
end
```

```rb
# prince
preflight do
  File.write shimscript, <<~EOS
    #!/bin/sh
    exec '#{staged_path}/prince-#{version}-macos/lib/prince/bin/prince' --prefix '#{staged_path}/prince-#{version}-macos/lib/prince' "$@"
  EOS
end
```

```rb
# subsync
preflight do
  File.write shimscript, <<~EOS
    #!/bin/sh
    exec '#{appdir}/subsync.app/Contents/MacOS/subsync' --cli "$@"
  EOS
end
```

**Comment by carlocab** [comment]
Looks fine to me then. @Homebrew/cask?

**Comment by SMillerDev** [comment]
Can we also check if the source exists and ensure the target directory does?

File: Library/Homebrew/cask/artifact/binary.rb [hunk 1]
```diff
@@ -3,14 +3,68 @@
...
+        args ||= {}
+
+        new(cask, source, **args)
+      end
+
+      def initialize(cask, source, target: nil, shimscript: nil)
+        super(cask, source, target: target)
+        return if shimscript.nil?
+
+        @shimscript_source = @source
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/artifact/binary.rb:33]:
```suggestion
        @shimscript_source = source
```

File: Library/Homebrew/cask/artifact/binary.rb [hunk 2]
```diff
@@ -3,14 +3,68 @@
...
+        if @shimscript_source
+          check_if_source_missing(@shimscript_source)
+
+          check_if_target_exists(source) do
+            next false unless force
+
+            source.realpath.to_s.start_with?("#{cask.caskroom_path}/")
+          end
+
+          ohai "Creating Shim Script of '#{@shimscript_source.basename}' at '#{source}'"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/artifact/binary.rb:48]:
```suggestion
          ohai "Creating shim script for '#{@shimscript_source.basename}' at '#{source}'"
```

File: Library/Homebrew/cask/artifact/binary.rb [hunk 3]
```diff
@@ -3,14 +3,68 @@
...
+        check_if_source_executable(source, command: command)
+        check_if_source_executable(@shimscript_source, command: command) if @shimscript_source
+      end
+
+      def unlink(**options)
+        super(**options)
+
+        return unless @shimscript_source
+
+        ohai "Removing Shim Script '#{source}'"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/artifact/binary.rb:63]:
```suggestion
        ohai "Removing shim script '#{source}'"
```

File: Library/Homebrew/cask/artifact/binary.rb [hunk 4]
```diff
@@ -3,14 +3,68 @@
...
+        super(force: force, command: command, **options)
+
+        check_if_source_executable(source, command: command)
+        check_if_source_executable(@shimscript_source, command: command) if @shimscript_source
+      end
+
+      def unlink(**options)
+        super(**options)
+
+        return unless @shimscript_source
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/artifact/binary.rb:61]:
```suggestion
        return if @shimscript_source.blank?
```

**Review by MikeMcQuaid — COMMENTED** [review]
Looking good so far!

File: Library/Homebrew/cask/artifact/binary.rb (hunk 1 above)
Comment (FnControlOption) [reply Library/Homebrew/cask/artifact/binary.rb]: This has to be `@source`, `self.source`, or `cask.staged_path.join(source) `

Here is where `@source` is set: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cask/artifact/relocated.rb#L53-L54

File: Library/Homebrew/cask/artifact/binary.rb [hunk 5]
```diff
@@ -3,14 +3,68 @@
...
     class Binary < Symlinked
-      def link(command: nil, **options)
-        super(command: command, **options)
+      def self.from_args(cask, *args)
+        source, args = args
+
+        if args
+          raise CaskInvalidError unless args.respond_to?(:keys)
+
+          args.assert_valid_keys!(:target, :shimscript)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/artifact/binary.rb:21]: Can we get some unit tests for the new bits?

**Pull request closed without merging**